File: dotnet/runtime/Models/GameState.cs
using System.Text.Json.Serialization;

namespace ColdChair.Runtime.Models;

public sealed class GameState
{
    [JsonPropertyName("tick")]
    public int Tick { get; init; }

    [JsonPropertyName("resources")]
    public required ResourceState Resources { get; init; }

    [JsonPropertyName("supply")]
    public required SupplyState Supply { get; init; }

    [JsonPropertyName("units")]
    public required List<UnitState> Units { get; init; }
}

public sealed class ResourceState
{
    [JsonPropertyName("gold")]
    public int Gold { get; init; }

    [JsonPropertyName("lumber")]
    public int Lumber { get; init; }
}

public sealed class SupplyState
{
    [JsonPropertyName("used")]
    public int Used { get; init; }

    [JsonPropertyName("cap")]
    public int Cap { get; init; }
}

public sealed class UnitState
{
    [JsonPropertyName("id")]
    public required string Id { get; init; }

    [JsonPropertyName("owner")]
    public required string Owner { get; init; }

    [JsonPropertyName("type")]
    public required string Type { get; init; }

    [JsonPropertyName("hp")]
    public double Hp { get; init; }

    [JsonPropertyName("position")]
    public required UnitPosition Position { get; init; }
}

public sealed class UnitPosition
{
    [JsonPropertyName("x")]
    public double X { get; init; }

    [JsonPropertyName("y")]
    public double Y { get; init; }
}

File: python/core/game_state.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class ResourceState:
    gold: int
    lumber: int


@dataclass
class SupplyState:
    used: int
    cap: int


@dataclass
class UnitState:
    id: str
    owner: str
    type: str
    hp: float
    x: float
    y: float


@dataclass
class GameState:
    tick: int
    resources: ResourceState
    supply: SupplyState
    units: list[UnitState]


def game_state_from_dict(payload: dict[str, Any]) -> GameState:
    resources_payload = payload.get("resources", {})
    supply_payload = payload.get("supply", {})
    units_payload = payload.get("units", [])

    units = [
        UnitState(
            id=str(unit["id"]),
            owner=str(unit["owner"]),
            type=str(unit["type"]),
            hp=float(unit["hp"]),
            x=float(unit["position"]["x"]),
            y=float(unit["position"]["y"]),
        )
        for unit in units_payload
    ]

    return GameState(
        tick=int(payload.get("tick", 0)),
        resources=ResourceState(
            gold=int(resources_payload.get("gold", 0)),
            lumber=int(resources_payload.get("lumber", 0)),
        ),
        supply=SupplyState(
            used=int(supply_payload.get("used", 0)),
            cap=int(supply_payload.get("cap", 0)),
        ),
        units=units,
    )
